Zotero/Scenes/Detail/PDF/ViewModels/PDFReaderActionHandler.swift

```diff
@@ -1396,32 +1374,170 @@ final class PDFReaderActionHandler: ViewModelActionHandler, BackgroundDbProcessi
     /// Updates annotations based on insertions to PSPDFKit document.
     /// - parameter annotations: Annotations that were added to the document.
     /// - parameter viewModel: ViewModel.
-    private func add(annotations: [PSPDFKit.Annotation], selectFirst: Bool, in viewModel: ViewModel<PDFReaderActionHandler>) {
-        guard let boundingBoxConverter = self.delegate else { return }
+    private func add(annotations: [PSPDFKit.Annotation], in viewModel: ViewModel<PDFReaderActionHandler>) {
+        guard let boundingBoxConverter = delegate else { return }
 
         DDLogInfo("PDFReaderActionHandler: annotations added - \(annotations.map({ "\(type(of: $0));key=\($0.key ?? "nil");" }))")
 
-        let finalAnnotations = self.splitIfNeededAndProcess(annotations: annotations, state: viewModel.state)
+        let (keptAsIs, toRemove, toAdd) = transformIfNeeded(annotations: annotations, state: viewModel.state)
+        let finalAnnotations = keptAsIs + toAdd
+        if !toRemove.isEmpty || !toAdd.isEmpty {
+            // Originally added annotations are transformed, so we remove them by performing last undo.
+            // This also removes the undo command from the stack, allowing us to record the transformed addition.
+            if viewModel.state.document.undoController.undoManager.canUndo {
```

Comment:
Looks good, but I worry a little, does this happen synchronously? When annotation is added to document, is this called immediately after? Couldn't another annotation be added before this is called so you'd undo different annotation? Is there a way to check that we're undo-ing the annotation we need? Or can we just remove the undo action manually since we're removing annotations below anyway?

Reply:
Did some improvements in https://github.com/zotero/zotero-ios/pull/934/commits/22f4389ed5c9d0d55e0f245b90900d7b76efbb22

It's immediatelly after the undo command is recorded. The sequence is as follows:
- add annotation programmatically in a recorded block
OR
add annotation via `AnnotationStateManager`, process notification with a debounce of `0` milliseconds so that the recording finishes
- `PDFReaderActionHandler` `add` method is called from the notification handler

AFAICS, we can either remove all undo actions, or call undo to remove the last one, and we cannot check programmatically what is in the undo/redo stacks. I researched other ways, like disabling automatic registration at document configuration and do all recordings programmatically, but the already drawn annotation when using the `AnnotationStateManager` causes a  lot of issues. 

We can ask PSPDFKit if they can add an option that passes starts detached recording, and passes the recorder with the notification, which will allow adding more commands to the recording block, and commiting afterwards. Still, there might be some issues I'm not aware of currently.

Reply:
Check https://github.com/zotero/zotero-ios/pull/934/commits/22546ac6f7301f41043768d30b9eef62a5755033